src/models/StyledComponent.js

```diff
@@ -235,20 +234,22 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
             ? rules
             : rulesFromOptions.concat(rules)
 
-        const newComponentId =
-          previousComponentId &&
-          `${previousComponentId}-${ComponentStyle.generateName(
-            newRules.join(''),
-          )}`
-
         const newOptions = {
           ...optionsToCopy,
-          componentId: newComponentId,
           rules: newRules,
           ParentComponent: StyledComponent,
         }
 
-        return constructWithOptions(createStyledComponent, target, newOptions)
+        const createWithHash = (tag, opts, css) => {
```

Comment:
@neoziro Since you're already fully at it and getting your hands dirty, I can guide you to what I mean :smile:

Here we could just pass in the `componentId` as a `parentComponentId`: https://github.com/styled-components/styled-components/blob/master/src/models/StyledComponent.js#L229

Then it'd eventually arrive back in the factory here: https://github.com/styled-components/styled-components/blob/master/src/models/StyledComponent.js#L178

There we can add a new argument to `generateId`:

```js
    const {
      componentId = generateId(options.displayName, options.parentComponentId)
    } = options
```

Then inside `generateId` you can concatenate this with `displayName` to make the new component id "more unique" i.e. different from the parent's one: https://github.com/styled-components/styled-components/blob/master/src/models/StyledComponent.js#L26

One thing we also need to do (**probably not in this PR**) is passing in the runtime inferred (default) `displayName` into `generateId` as well instead of just `options.displayName`: https://github.com/styled-components/styled-components/blob/master/src/models/StyledComponent.js#L177